<h1 id="buckley-lab-sip-protocols">Buckley Lab SIP protocols</h1>
<h2 id="printing-protocols-conversion-of-protocols-to-pdf">Printing protocols (conversion of protocols to PDF)</h2>
<p>View any Markdown file on GitHub, then in your URL bar replace the git<strong>hub</strong>.com part of the URL with git<strong>print</strong>.com</p>
<p>The Markdown file will be rendered as a PDF for easy printing or downloading.</p>
<h2 id="updating-non-markdown-versions-of-protocol-files">Updating non-markdown versions of protocol files</h2>
<ul>
<li>Please don't edit the Word or html files directly!</li>
<li>The markdown files (*.md) serve as the template files for all of the other formats (eg., html or docx).</li>
<li>The bash scripts (eg., md2html.sh) will convert all files ending in '.md' to a specified format.
<ul>
<li>Example usage: <code>./md2html.sh</code>
<ul>
<li>This will creat a html file for each corresponding markdown file.</li>
</ul></li>
</ul></li>
</ul>
<h2 id="working-in-the-lab">Working in the lab</h2>
<ul>
<li><a href="./working_in_the_lab/lab_etiquette.md">lab_etiquette</a></li>
<li><a href="./working_in_the_lab/undergrad_expectations.md">undergrad_expectations</a></li>
</ul>
<h2 id="pipeline-steps">Pipeline steps</h2>
<ul>
<li>Soil sampling
<ul>
<li><a href="./sampling/soil_sampling_protocol.md">soil_sampling_protocol</a></li>
</ul></li>
<li>Microcosm setup
<ul>
<li>Ashley's priming experiment setup
<ul>
<li><a href="./microcosm/SIP_microcosm.md">SIP_microcosm</a></li>
</ul></li>
<li>Nick's full cycle pilot exp. setup
<ul>
<li><a href="./microcosm/SIP_fullCyc_pilot_microcosm.md">SIP_fullCyc_pilot_microcosm</a></li>
</ul></li>
</ul></li>
<li>Microcosm headspace CO<sub>2</sub> measurements (via GC/MS)
<ul>
<li><a href="./GCMS_operation/CO2_batch_run.md">CO2_batch_run</a></li>
</ul></li>
<li>Nucleotide extraction
<ul>
<li><a href="./nucleotide_extraction/DNA_RNA_extraction_Protocol.md">DNA_RNA_extraction_Protocol</a></li>
</ul></li>
<li><strong>If DNA:</strong> CsCl fractionation
<ul>
<li><a href="./CsCl_fractionation/CsCl_fractionation.md">CsCl_fractionation</a></li>
</ul></li>
<li><strong>Else if RNA:</strong> CsTFA fractionation
<ul>
<li><a href="./RNA_SIP/RNA_SIP.md">RNA_SIP</a></li>
</ul></li>
<li>Nucleotide quantification:
<ul>
<li><a href="./nucleotide_conc/picogreen.md">Picogreen</a></li>
</ul></li>
<li>Nucleotide sample concentration
<ul>
<li><a href="./speed-vac/speed-vac.md">speed-vac</a></li>
</ul></li>
<li>Pippin Prep
<ul>
<li><a href="./Pippin_prep/Pippin_prep.md">Pippin_prep</a></li>
</ul></li>
<li>Fraction nucleotide quantification:
<ul>
<li><a href="./nucleotide_conc/picogreen.md">picogreen</a></li>
</ul></li>
<li>MiSeq library prep:
<ul>
<li><a href="./library_prep/Illumina_barcoding_protocol.md">Illumina_barcoding_protocol</a></li>
</ul></li>
</ul>
<h2 id="others">Others</h2>
<ul>
<li>Cellulose farming
<ul>
<li>Cellulose production
<ul>
<li><a href="./cellulose_farming/CelluloseProductionProtocol.md">CelluloseProductionProtocol</a></li>
</ul></li>
<li>Cellulose grinding:
<ul>
<li><a href="./cellulose_farming/CelluloseGrindingProtocol.md">CelluloseGrindingProtocol</a></li>
</ul></li>
</ul></li>
<li>Plant stimulant
<ul>
<li><a href="./plant_stimulant/Substrate_Additions_MicrobSuccession.md">Substrate_Additions_MicrobSuccession</a></li>
</ul></li>
<li>Soil geochemistry
<ul>
<li><a href="./soil_geochemistry/pH.md">pH</a></li>
<li><a href="./soil_geochemistry/organic_content.md">organic_content</a></li>
<li><a href="./soil_geochemistry/water_holding_capacity.md">water_holding_capacity</a></li>
</ul></li>
</ul>
<h2 id="workflows">Workflows</h2>
<h3 id="bulk-dna-sequencing-of-the-16s-rrna-gene">Bulk DNA sequencing of the 16S rRNA gene</h3>
<ul>
<li><p><a href="./nucleotide_extraction/DNA_RNA_extraction_Protocol.md">Nucleotide extraction</a></p></li>
<li><p><a href="http://www.gelifesciences.com/webapp/wcs/stores/servlet/productById/en/GELifeSciences/27533001">Sephadex column clean-up</a></p></li>
<li><p><a href="./nucleotide_conc/picogreen.md">Nucleotide quantification via Picogreen</a></p></li>
<li><p><a href="./library_prep/Illumina_barcoding_protocol.md">16S rRNA amplicon library prep</a></p></li>
</ul>
<h3 id="microcosm---gradient-fractionation---16s-rrna-gene-sequencing">Microcosm -&gt; gradient fractionation -&gt; 16S rRNA gene sequencing</h3>
<ul>
<li><p><a href="./nucleotide_extraction/DNA_RNA_extraction_Protocol.md">Nucleotide extraction</a></p></li>
<li><p><strong>Optional:</strong> Concentrating via <a href="./speed-vac/speed-vac.md">speed-vac</a></p></li>
<li><p><a href="./Pippin_prep/Pippin_prep.md">PippinPrep</a></p></li>
<li><p><a href="./nucleotide_conc/picogreen.md">Nucleotide quantification via Picogreen</a></p></li>
<li><a href="./CsCl_fractionation/CsCl_fractionation.md">CsCl_fractionation</a>
<ul>
<li>Including desalting</li>
</ul></li>
<li><p><a href="./nucleotide_conc/picogreen.md">Nucleotide quantification of fractions via Picogreen</a></p></li>
<li><a href="./library_prep/Illumina_barcoding_protocol.md">16S rRNA amplicon library prep</a>
<ul>
<li>For automated generation of the Excel files needed for the PCR assay, see the fractionSelectFor16S-PCR.ipynb notebook in the SIPdb repo.</li>
</ul></li>
</ul>
