{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> ### Pipeling to Merge Sequencing Libraries from Different Sequencing Runs that have been denoised by DADA2  and Finish with the Standard Processing Pipeline###\n",
    "* Merge multiple libraries for the *SAME* sequencing region (i.e. ITS+ITS... or SSU+SSU...)\n",
    "* Complete remainder of processing pipieline: prepare Rep Seqs, Classify Seqs, Make Tree, Make Phyloseq Object\n",
    "\n",
    "*Informed by the \"Fecal Microbiota Transplant Tutorial\" for QIIME2*\n",
    "\n",
    "### Commands to Install Dependencies ####\n",
    "##### || QIIME2 ||\n",
    "* conda create -n qiime2-2018.2 --file https://data.qiime2.org/distro/core/qiime2-2018.2-conda-linux-64.txt  \n",
    "\n",
    "* source activate qiime2-2018.2\n",
    "\n",
    "##### || Copyrighter rrn Database ||\n",
    "* The script will automatically install the curated GreenGenes rrn attribute database\n",
    "* https://github.com/fangly/AmpliCopyrighter\n",
    "\n",
    "##### || rpy2 (don't use conda version) ||\n",
    "* pip install rpy2  \n",
    "\n",
    "##### || phyloseq ||\n",
    "* conda install -c r-igraph \n",
    "* Rscript -e \"source('http://bioconductor.org/biocLite.R');biocLite('phyloseq')\" \n",
    "\n",
    "##### || R packages ||\n",
    "* ape   (natively installed with in conda environment)\n",
    "\n",
    "\n",
    "### Citations ###\n",
    "* Caporaso, J. G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F. D., Costello, E. K., *et al.* (2010). QIIME allows analysis of high-throughput community sequencing data. Nature methods, 7(5), 335-336.\n",
    "\n",
    "\n",
    "* McMurdie and Holmes (2013) phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data. PLoS ONE. 8(4):e61217\n",
    "\n",
    "\n",
    "* Paradis E., Claude J. & Strimmer K. 2004. APE: analyses of phylogenetics and evolution in R language. Bioinformatics 20: 289-290.\n",
    "\n",
    "\n",
    "* Angly, F. E., Dennis, P. G., Skarshewski, A., Vanwonterghem, I., Hugenholtz, P., & Tyson, G. W. (2014). CopyRighter: a rapid tool for improving the accuracy of microbial community profiles through lineage-specific gene copy number correction. Microbiome, 2(1), 11.\n",
    "\n",
    "\n",
    "###### Last Modified by R. Wilhelm on October 12th, 2017 ######\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: User Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re\n",
    "\n",
    "# Provide the directory where files are located \n",
    "#directory = '/home/user/PROJECT/16S/'\n",
    "directory = ''\n",
    "\n",
    "# Provide a list of all the FeatureTables you will merge\n",
    "# Produced by QIIME2 in STEP 7 (i.e. DADA2 Denoising/Merging/FeatureTable)\n",
    "#qva_files = ['SSU.table.Library.1.qza','SSU.table.Library.2.qza', ..., 'SSU.table.Library.n.qza']\n",
    "qva_files = ['','','']\n",
    "\n",
    "# Provide a list of all the Representative Sequences you will merge\n",
    "# Also produced from STEP 7 (i.e. DADA2 Denoising/Merging/FeatureTable)\n",
    "#seq_files = ['SSU.rep.seqs.Library.1.qza','SSU.rep.seqs.Library.2.qza', ..., 'SSU.rep.seqs.Library.n.qza']\n",
    "seq_files = ['','','']\n",
    "\n",
    "# Provide a concatenated metadatafile with matching column order\n",
    "#metadata = 'SSU.metadata.combined.tsv'\n",
    "metadata = ''\n",
    "\n",
    "## Enter Minimum Support for Keeping QIIME Classification\n",
    "# Note: Classifications that do not meet this criteria will simply be retained, but labeled 'putative'\n",
    "min_support = 0.8\n",
    "\n",
    "domain = 'bacteria'   # 'bacteria' | 'fungi'\n",
    "\n",
    "\n",
    "##################################    IMPORTANT    #####################################\n",
    "##  DO NOT give your samples names that begin with a digit. It will break this script ##\n",
    "##  at the import to phyloseq stage. It doesn't ruin anything, but makes it a pain.   ##\n",
    "########################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Merge Feature Tables using QIIME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%mkdir $directory/output\n",
    "\n",
    "combined_qva = [m+\" \"+n for m,n in zip([\"--i-tables\"]*len(qva_files),qva_files)]\n",
    "\n",
    "os.system(' '.join([\n",
    "    \"qiime feature-table merge\",\n",
    "    ' '.join(combined_qva),\n",
    "    \"--o-merged-table \"+directory+\"/output/merged.table.final.qza\"\n",
    "]))\n",
    "\n",
    "\n",
    "combined_seq = [m+\" \"+n for m,n in zip([\"--i-data\"]*len(seq_files),seq_files)]\n",
    "\n",
    "os.system(' '.join([\n",
    "    \"qiime feature-table merge\",\n",
    "    ' '.join(combined_seq),\n",
    "    \"--o-merged-table \"+directory+\"/output/merged.table.final.qza\"\n",
    "]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Create Summary of OTUs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(' '.join([\n",
    "    \"qiime feature-table summarize\",\n",
    "    \"--i-table \"+directory+\"/output/merged.table.final.qza\",\n",
    "    \"--o-visualization \"+directory+\"/output/merged.table.final.qzv\",\n",
    "    \"--m-sample-metadata-file \"+directory+metadata\n",
    "]))\n",
    "\n",
    "os.system(' '.join([\n",
    "    \"qiime feature-table tabulate-seqs\",\n",
    "    \"--i-data \"+directory+\"/output/merged.rep.seqs.final.qza\",\n",
    "    \"--o-visualization \"+directory+\"/output/merged.rep.seqs.final.qzv\"\n",
    "])) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 4: Make Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if domain != \"fungi\":\n",
    "    # Generate Alignment with MAFFT\n",
    "    os.system(' '.join([\n",
    "        \"qiime alignment mafft\",\n",
    "        \"--i-sequences \"+directory+\"/output/merged.rep.seqs.final.qza\",\n",
    "        \"--o-alignment \"+directory+\"/output/merged.rep.seqs.aligned.qza\"\n",
    "    ]))\n",
    "\n",
    "    # Mask Hypervariable parts of Alignment\n",
    "    os.system(' '.join([\n",
    "        \"qiime alignment mask\",\n",
    "        \"--i-alignment \"+directory+\"/output/merged.rep.seqs.aligned.qza\",\n",
    "        \"--o-masked-alignment \"+directory+\"/output/merged.rep.seqs.aligned.masked.qza\"\n",
    "    ])) \n",
    "\n",
    "    # Generate Tree with FastTree\n",
    "    os.system(' '.join([\n",
    "        \"qiime phylogeny fasttree\",\n",
    "        \"--i-alignment \"+directory+\"/output/merged.rep.seqs.aligned.masked.qza\",\n",
    "        \"--o-tree \"+directory+\"/output/merged.rep.seqs.tree.unrooted.qza\"\n",
    "    ])) \n",
    "\n",
    "    # Root Tree\n",
    "    os.system(' '.join([\n",
    "        \"qiime phylogeny midpoint-root\",\n",
    "        \"--i-tree \"+directory+\"/output/merged.rep.seqs.tree.unrooted.qza\",\n",
    "        \"--o-rooted-tree \"+directory+\"/output/merged.rep.seqs.tree.final.qza\"\n",
    "    ])) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5: Classify Seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Classify\n",
    "if domain == 'bacteria':\n",
    "    os.system(' '.join([\n",
    "        \"qiime feature-classifier classify-sklearn\",\n",
    "        \"--i-classifier /home/db/GreenGenes/qiime2_13.8.99_515.806_nb.classifier.qza\",\n",
    "        \"--i-reads \"+directory+\"/output/merged.rep.seqs.final.qza\",\n",
    "        \"--o-classification \"+directory+\"/output/merged.taxonomy.final.qza\"\n",
    "    ]))\n",
    "\n",
    "if domain == 'fungi':\n",
    "    os.system(' '.join([\n",
    "        \"qiime feature-classifier classify-sklearn\",\n",
    "        \"--i-classifier /home/db/UNITE/qiime2_unite_ver7.99_20.11.2016_classifier.qza\",\n",
    "        \"--i-reads \"+directory+\"/output/merged.rep.seqs.final.qza\",\n",
    "        \"--o-classification \"+directory+\"/output/merged.taxonomy.final.qza\"\n",
    "    ]))\n",
    "\n",
    "# Output Summary\n",
    "os.system(' '.join([\n",
    "    \"qiime metadata tabulate\",\n",
    "    \"--m-input-file \"+directory+\"/output/merged.taxonomy.final.qza\",\n",
    "    \"--o-visualization \"+directory+\"/output/merged.taxonomy.final.summary.qzv\"\n",
    "])) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 6: Prepare Data for Import to Phyloseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make Function to Re-Format Taxonomy File to Contain Full Column Information \n",
    "# and factor in the certain of the taxonomic assignment\n",
    "\n",
    "def format_taxonomy(tax_file, min_support):\n",
    "    output = open(re.sub(\".tsv\",\".fixed.tsv\",tax_file), \"w\")\n",
    "    output.write(\"\\t\".join([\"OTU\",\"Domain\",\"Phylum\",\"Class\",\"Order\",\"Family\",\"Genus\",\"Species\"])+\"\\n\")\n",
    "    \n",
    "    with open(tax_file, \"r\") as f:\n",
    "        next(f) #skip header\n",
    "\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            line = line.split(\"\\t\")\n",
    "\n",
    "            read_id = line[0]\n",
    "            tax_string = line[1]\n",
    "\n",
    "            # Annotate those strings which do not meet minimum support\n",
    "            if float(line[2]) < float(min_support):\n",
    "                tax_string = re.sub(\"__\",\"__putative \",tax_string)\n",
    "\n",
    "            # Remove All Underscore Garbage (gimmie aesthetics)\n",
    "            tax_string = re.sub(\"k__|p__|c__|o__|f__|g__|s__\",\"\",tax_string) \n",
    "\n",
    "            # Add in columns containing unclassified taxonomic information\n",
    "            # Predicated on maximum 7 ranks (Domain -> Species)\n",
    "            full_rank = tax_string.split(\";\")\n",
    "            last_classified = full_rank[len(full_rank)-1]\n",
    "\n",
    "            count = 1\n",
    "            while last_classified == \" \":\n",
    "                last_classified = full_rank[len(full_rank)-count]\n",
    "                count = count + 1\n",
    "\n",
    "\n",
    "            for n in range(full_rank.index(last_classified)+1, 7, 1):\n",
    "                try:\n",
    "                    full_rank[n] = \"unclassified \"+last_classified\n",
    "                except:\n",
    "                    full_rank.append(\"unclassified \"+last_classified)\n",
    "\n",
    "            output.write(read_id+\"\\t\"+'\\t'.join(full_rank)+\"\\n\")\n",
    "            \n",
    "    return()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#####################\n",
    "## Export from QIIME2\n",
    "\n",
    "## Final Output Names\n",
    "fasta_file = directory+\"/output/merged.rep.seqs.final.fasta\"\n",
    "tree_file = directory+\"/output/merged.tree.final.nwk\"\n",
    "tax_file = directory+\"/output/merged.taxonomy.final.tsv\"\n",
    "count_table = directory+\"/output/merged.counts.final.biom\"\n",
    "\n",
    "# Export Classifications\n",
    "os.system(' '.join([\n",
    "    \"qiime tools export\",\n",
    "    directory+\"/output/merged.taxonomy.final.qza\",\n",
    "    \"--output-dir \"+directory+\"/output/\"\n",
    "]))\n",
    "\n",
    "# Reformat Classifications to meet phyloseq format\n",
    "format_taxonomy(directory+\"/output/taxonomy.tsv\", min_support)\n",
    "\n",
    "# Export SV Table\n",
    "os.system(' '.join([\n",
    "    \"qiime tools export\",\n",
    "    directory+\"/output/merged.table.final.qza\",\n",
    "    \"--output-dir \"+directory+\"/output/\"\n",
    "]))\n",
    "\n",
    "# Export SV Sequences\n",
    "os.system(' '.join([\n",
    "    \"qiime tools export\",\n",
    "    directory+\"/output/merged.rep.seqs.final.qza\",\n",
    "    \"--output-dir \"+directory+\"/output/\"\n",
    "]))\n",
    "\n",
    "if domain == \"bacteria\":\n",
    "    # Export Tree\n",
    "    os.system(' '.join([\n",
    "        \"qiime tools export\",\n",
    "        directory+\"/output/merged.rep.seqs.tree.final.qza\",\n",
    "        \"--output-dir \"+directory+\"/output/\"\n",
    "    ]))\n",
    "\n",
    "    %mv $directory/output/tree.nwk $tree_file\n",
    "\n",
    "# Rename Exported Files\n",
    "%mv $directory/output/dna-sequences.fasta $fasta_file\n",
    "%mv $directory/output/feature-table.biom $count_table\n",
    "%mv $directory/output/taxonomy.fixed.tsv $tax_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 7: Get 16S rRNA Gene Copy Number (rrn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This step is based on the database contructed for the software 'copyrighter'\n",
    "## The software itself lacked information about datastructure (and, the import of a biom from QIIME2 failed, likely because there are multiple versions of the biom format)\n",
    "\n",
    "if domain == 'bacteria':\n",
    "    ## Download copyrighter database\n",
    "    !git clone https://github.com/fangly/AmpliCopyrighter $directory/temp/\n",
    "        \n",
    "    ## There are multiple GreenGenes ID numbers for a given taxonomic string.\n",
    "    ## However, the copyrighter database uses the same average rrn copy number.\n",
    "    ## We will therefore just use the taxonomic strings, since QIIME2 does not output the ID numbers\n",
    "\n",
    "    !sed -e '1,1075178d; 1078115d' $directory/temp/data/201210/ssu_img40_gg201210.txt > $directory/output/copyrighter.tax.strings.tsv\n",
    "\n",
    "    ## Create Dictionary of rrnDB\n",
    "    rrnDB = {}\n",
    "\n",
    "    with open(directory+\"/output/copyrighter.tax.strings.tsv\", \"r\") as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            line = line.split(\"\\t\")\n",
    "\n",
    "            try:\n",
    "                rrnDB[line[0]] = line[1]\n",
    "                \n",
    "            except:\n",
    "                pass\n",
    "\n",
    "    ## Attribute rrn to readID from taxonomy.tsv\n",
    "    output = open(directory+\"/output/merged.seqID.to.rrn.final.tsv\",\"w\")\n",
    "    output.write(\"Feature ID\\trrn\\n\")\n",
    "\n",
    "    with open(directory+\"/output/taxonomy.tsv\", \"r\") as f:\n",
    "        missing = 0\n",
    "        total = 0\n",
    "        next(f)  # Skip Header\n",
    "\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            line = line.split(\"\\t\")\n",
    "\n",
    "            seqID = line[0]\n",
    "\n",
    "            try:\n",
    "                rrn = rrnDB[line[1]]\n",
    "                \n",
    "            except:\n",
    "                rrn = \"NA\"\n",
    "                missing = missing + 1\n",
    "\n",
    "            total = total + 1\n",
    "            output.write(seqID+\"\\t\"+rrn+\"\\n\")\n",
    "\n",
    "    print(\"\\nPercent of OTUs Missing {:.1%}\".format(float(missing)/total))\n",
    "    print(\"Don't Panic! The majority of missing OTUs could be low abundance.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 8: Import into Phyloseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Setup R-Magic for Jupyter Notebooks\n",
    "import rpy2\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "def fix_biom_conversion(file):\n",
    "    with open(file, 'r') as fin:\n",
    "        data = fin.read().splitlines(True)\n",
    "    with open(file, 'w') as fout:\n",
    "        fout.writelines(data[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "%R library(phyloseq)\n",
    "%R library(ape)\n",
    "\n",
    "#### IMPORT DATA to R\n",
    "## For '.tsv' files, use Pandas to create a dataframe and then pipe that to R\n",
    "## For '.biom' files, first convert using 'biom convert' on the command-line\n",
    "## Had problems importing the count table with pandas, opted for using read.table in R\n",
    "\n",
    "# Import Taxonomy File\n",
    "tax_file = pd.read_csv(directory+\"/output/merged.taxonomy.final.tsv\", sep=\"\\t\")\n",
    "%R -i tax_file\n",
    "%R rownames(tax_file) = tax_file$OTU\n",
    "%R tax_file$OTU <- NULL\n",
    "%R tax_file <- tax_file[sort(row.names(tax_file)),] #read names must match the count_table\n",
    "\n",
    "# Import Sample Data\n",
    "sample_file = pd.read_csv(directory+metadata, sep=\"\\t\", keep_default_na=False)\n",
    "%R -i sample_file\n",
    "%R rownames(sample_file) = sample_file$X.SampleID   \n",
    "%R sample_file$X.SampleID <- NULL\n",
    "%R sample_file$LinkerPrimerSequence <- NULL  ## Clean-up some other stuff\n",
    "\n",
    "# Import Count Data\n",
    "os.system(' '.join([\n",
    "    \"biom convert\",\n",
    "    \"-i\",\n",
    "    directory+\"/output/merged.counts.final.biom\",\n",
    "    \"-o\",\n",
    "    directory+\"/output/merged.counts.final.tsv\",\n",
    "    \"--to-tsv\"\n",
    "]))\n",
    "\n",
    "# The biom converter adds a stupid line that messes with the table formatting\n",
    "fix_biom_conversion(directory+\"/output/merged.counts.final.tsv\")\n",
    "\n",
    "# Finally import\n",
    "count_table = pd.read_csv(directory+\"/output/merged.counts.final.tsv\", sep=\"\\t\")\n",
    "%R -i count_table\n",
    "%R rownames(count_table) = count_table$X.OTU.ID   \n",
    "%R count_table$X.OTU.ID <- NULL    \n",
    "%R count_table <- count_table[sort(row.names(count_table)),] #read names must match the tax_table\n",
    "\n",
    "# Convert to Phyloseq Objects\n",
    "%R p_counts = otu_table(count_table, taxa_are_rows = TRUE)    \n",
    "%R p_samples = sample_data(sample_file)    \n",
    "%R p_tax = tax_table(tax_file)\n",
    "%R taxa_names(p_tax) <- rownames(tax_file) # phyloseq throws out rownames\n",
    "%R colnames(p_tax) <- colnames(tax_file) # phyloseq throws out colnames\n",
    "\n",
    "# Merge Phyloseq Objects\n",
    "%R p = phyloseq(p_counts, p_tax)\n",
    "\n",
    "if domain == \"bacteria\":\n",
    "    # Import Phylogenetic Tree\n",
    "    tree_file = directory+\"/output/merged.tree.final.nwk\"\n",
    "    %R -i tree_file  \n",
    "    %R p_tree <- read.tree(tree_file)\n",
    "\n",
    "    # Combine All Objects into One Phyloseq\n",
    "    %R p_final <- merge_phyloseq(p, p_samples, p_tree)\n",
    "\n",
    "else:\n",
    "    # Combine All Objects into One Phyloseq\n",
    "    %R p_final <- merge_phyloseq(p, p_samples)\n",
    "    \n",
    "# Save Phyloseq Object as '.rds'\n",
    "output = directory+\"/output/p_merged.final.rds\"\n",
    "%R -i output\n",
    "%R saveRDS(p_final, file = output)\n",
    "\n",
    "# Confirm Output\n",
    "%R print(p_final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 9: Clean-up Intermediate Files and Final Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You'll have to delete the 'temp' directory manually.\n",
    "\n",
    "# Remove Files\n",
    "if domain == \"bacteria\":\n",
    "    %rm $directory/output/*tree.unrooted.qza \n",
    "    %rm $directory/output/*aligned.masked.qza \n",
    "    \n",
    "%rm $directory/output/*.biom \n",
    "%rm $directory/output/taxonomy.tsv\n",
    "%rm $directory/output/copyrighter.tax.strings.tsv\n",
    "\n",
    "# Separate Final Files\n",
    "%mkdir $directory/final/    \n",
    "%mv $directory/output/*.final.rds $directory/final/\n",
    "%mv $directory/output/*.taxonomy.final.tsv $directory/final/    \n",
    "%mv $directory/output/*.counts.final.tsv $directory/final/\n",
    "%mv $directory/output/*.final.fasta $directory/final/\n",
    "%cp $directory$metadata $directory/final/\n",
    "%mv $directory/output/merged.seqID.to.rrn.final.tsv $directory/final/\n",
    "\n",
    "# Gzip and Move Intermediate Files\n",
    "!pigz -p 10 $directory/output/*.qza\n",
    "!pigz -p 10 $directory/output/*.qzv\n",
    "%mv $directory/output/ $directory/intermediate_files\n",
    "\n",
    "print(\"Your sequences have been successfully saved to directories 'final' and 'intermediate_files'\")"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "hide_input": true,
  "kernelspec": {
   "display_name": "Environment (conda_qiime2-2017.9)",
   "language": "python",
   "name": "conda_qiime2-2017.9"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
