<h1 id="illumina-miseq-barcoding-protocol">Illumina Miseq Barcoding Protocol</h1>
<h2 id="authorship">Authorship</h2>
<p>Ashley Campbell, Chuck Pepe-Ranney, Chantal Koechli, and Nick Youngblut (2014)</p>
<h1 id="method">Method</h1>
<ol style="list-style-type: decimal">
<li><p>Quantify samples to be sequenced using Picogreen protocol.</p></li>
<li><p>Calculate sample volume necessary for addition of <strong>5 ng</strong> of template DNA to PCR reactions.</p></li>
<li><p>PCR reaction composition for 1 rxn:</p>
<ul>
<li>12.5 uL Mastermix (NEB Q5 High Fidelity, Hot Start PCR Mastermix - M0494)
<ul>
<li>50% of total volume</li>
</ul></li>
<li>2.5 uL combined forward and reverse barcoded primers (10 uM)
<ul>
<li>10% of total volume<br /></li>
<li>Primer plates can be created with the robot method: &quot;make_primer_plate&quot;</li>
</ul></li>
<li>1.25 uL BSA (20 mg/mL, NEB B9000S)
<ul>
<li>5% of total volume</li>
</ul></li>
<li>0.625 uL Picogreen reagent
<ul>
<li>2% of total volume</li>
<li>4x concentration, dilute the 200x stock that comes in the Picogreen kit with 1X TE</li>
</ul></li>
<li><p>X uL template (5 ng/reaction)</p></li>
<li><p>PCR water up to 25 uL</p></li>
</ul></li>
</ol>
<ul>
<li><strong>If adding pico &amp; BSA to master mix prior to aliquoting master mix:</strong>
<ul>
<li>total_volume = master_mix + BSA + pico</li>
<li>volume_master_mix = total_volume * 0.870</li>
<li>volume_BSA = total_volume * 0.087</li>
<li>volume_pico = total_volume * 0.043</li>
</ul></li>
</ul>
<ol style="list-style-type: decimal">
<li><p>Use robot method &quot;qPCR_wWorklist_altdispense&quot; for setting up PCRs, running triplicate reactions for each sample to be sequenced.</p></li>
<li>Run the PCR plate on the qPCR thermocycler, using the following cycle:
<ul>
<li>98<sup>o</sup>C for 30 seconds</li>
<li>30 cycles of:
<ul>
<li>98<sup>o</sup>C for 5 seconds</li>
<li>50<sup>o</sup>C for 30 seconds</li>
<li>72<sup>o</sup>C for 10 seconds</li>
</ul></li>
<li>Final extension of 72<sup>o</sup>C for 2 minutes</li>
<li>Hold at 4<sup>o</sup>C</li>
</ul></li>
<li>Combine triplicate PCRs for each samples, transferring samples to a new 96-well plate.
<ul>
<li>This can be done with the robot method: &quot;plate_pooling&quot;</li>
</ul></li>
<li>Perform Sequal PCR purification and normalization
<ul>
<li>SequalPrep Normalization Plates, Life Technologies, A10510-01</li>
<li>Follow the <a href="https://www.lifetechnologies.com/order/catalog/product/A1051001">manufacturer's instructions</a>, using 25 uL of PCR product for each sample.</li>
<li>This can be done with the robot method: &quot;SequalPrep_Assay&quot;</li>
</ul></li>
<li>Combine all Sequal'd samples (20 ul/sample) into one tube (or two, if the volume too large).
<ul>
<li>Pre-weigh the tube to help with the next step (speed-vac).</li>
<li>This can be done with the robot method: &quot;plate_pooling&quot;</li>
</ul></li>
<li>Vacuum evaporate samples to concentrate.
<ul>
<li><a href="../speed-vac/speed-vac.html">speed-vac</a></li>
<li>You need &gt;=25 uL with a concentration of 5 ng/uL</li>
</ul></li>
<li>Quantify concentrated collection of samples using PicoGreen.
<ul>
<li>Make sure to have &gt;=25 uL with a concentration of 5 ng/uL.</li>
</ul></li>
<li>Run concentrated sample on a gel to ensure expected, cleaned product for sequencing.
<ul>
<li>If additional, unexpected bands are seen, consider <a href="../gel_extraction/gel_extraction.html">Gel extraction</a> of the concentrated sample.
<ul>
<li>This may result in higher sequence quality.</li>
</ul></li>
</ul></li>
<li>Submission to Cornell Sequencing Facility:
<ul>
<li>Place 20 uL of sample with a concentration of 5 ng/uL into sequencing tube.</li>
<li>Additionally, submit 10 uL of 100 uM sequencing primers:
<ul>
<li>forward sequencing primer</li>
<li>reverse sequencing primer</li>
<li>reverse index read primer</li>
</ul></li>
</ul></li>
</ol>
<h2 id="usual-comments-provided-with-the-sequencing-order">Usual comments provided with the sequencing order:</h2>
<h3 id="comments-for-515f-806r-primers">Comments for 515f-806r primers</h3>
<p>We have generated our dual-indexed custom barcoded library as described in Kozich et al., (2013). Along with the submitted library, we have provided 10 ul of 100 uM custom sequencing primers: Primer 1, Index Primer, Primer 2.</p>
<p>We request the following MiSeq run specifications: a <strong>cluster density of 650-750k/mm^2</strong> (under-shooting the cluster density is better than over-shooting), a <strong>PhiX spike-in of 5%</strong>. These MiSeq run specifications are described in Kozich et al., (2013); see 'Run Monitoring' in the Supplemental Materials.</p>
<p>In addition, we have been in contact with Peter Schweitzer on using dPCR to for library quantification in hopes of obtaining more accurate cluster densities. The dPCR assay entails using the ABI QuantStudio3D instrument with SYBR Green and primers that target the Illumina adaptor sequences. The data that Peter has shared with us looks promising, but taking the quantifications at face value would likely cause <strong>over-clustering</strong>, and thus the raw values likely need to be corrected in order to achieve the target cluster density. Please let us know if dPCR cannot be used to quantify our libraries or if you have any other questions about the assay.</p>
<p>Also, please let us know if you have any questions about the primers or requested run parameters. Thank you!</p>
<p>[Citation: Kozich JJ, Westcott SL, Baxter NT, Highlander SK, Schloss PD. (2013). Development of a Dual-Index Sequencing Strategy and Curation Pipeline for Analyzing Amplicon Sequence Data on the MiSeq Illumina Sequencing Platform. Appl Environ Microbiol 79:5112-5120.]</p>
<h3 id="comments-for-515f-927r-primers">Comments for 515f-927r primers</h3>
<blockquote>
<p>We would like to run the 2x300 paired end read platform. Our submitted sample is a pooled set of already barcoded samples. We have supplied custom primers (labeled Primer 1, Index Primer 1, and Primer 2) for the sequencing as well, as follows: Primer 1: combined pad, linker, and gene-specific primer sequence at the 5' end of the region (pad+linker+515F primer) that will produce a long sequence read from the 5' end of the sequence. (Read 1) Index Primer 1: reverse complement of the pad, linker, and gene-specific primer from the 3'end (pad+linker+927R) that will provide an index read of the 3' end of the sequence. (Read 2) Primer 2: Combined pad linker, and gene-specific primer sequence at the 3' end of the region (pad+linker+927R primer) that will produce the second sequence read (300 nt) from the 3' end of the sequence (Read 4). The primers are described in more detail in [Kozich JJ, Westcott SL, Baxter NT, Highlander SK, Schloss PD. (2013). Development of a Dual-Index Sequencing Strategy and Curation Pipeline for Analyzing Amplicon Sequence Data on the MiSeq Illumina Sequencing Platform. Appl Environ Microbiol 79:5112-5120]. 10ul of 100uM primer has been supplied. Please let us know if you have any questions about the primers.</p>
</blockquote>
